src/pb_stub.cc

```diff
@@ -909,29 +930,186 @@ Stub::SendLogMessage(std::unique_ptr<PbLog>& log_send_message)
+  cleanup_message_ptr->id = id;
+  ipc_message->Args() = cleanup_request_message.handle_;
+
+  ScopedDefer data_load_complete([&ipc_message] {
+    bi::scoped_lock<bi::interprocess_mutex> lock{
+        *(ipc_message->ResponseMutex())};
+    ipc_message->ResponseCondition()->notify_all();
+  });
+
+  {
+    bi::scoped_lock<bi::interprocess_mutex> lock{
+        *(ipc_message->ResponseMutex())};
+    SendIPCUtilsMessage(ipc_message);
+    ipc_message->ResponseCondition()->wait(lock);
+  }
+}
+
+void
+Stub::EnqueueCleanupId(void* id)
+{
+  {
+    std::lock_guard<std::mutex> guard{utils_message_mutex_};
+    bls_response_cleanup_buffer_.push(id);
+  }
+  utils_message_cv_.notify_one();
+}
+
 bool
-Stub::LogServiceActive()
+Stub::UtilsServiceActive()
 {
-  return log_thread_;
+  return utils_thread_;
+}
+
+void
+Stub::LaunchBLSResponseQueueMonitor()
+{
+  bls_response_thread_ = true;
+  bls_response_monitor_ = std::thread(&Stub::BLSResponseQueueMonitor, this);
+}
+
+void
+Stub::TerminateBLSResponseQueueMonitor()
+{
+  if (bls_response_thread_) {
+    bls_response_thread_ = false;
+    bls_response_queue_->Push(DUMMY_MESSAGE);
+    bls_response_monitor_.join();
+  }
+}
+
+void
+Stub::BLSResponseQueueMonitor()
+{
+  while (bls_response_thread_) {
+    bi::managed_external_buffer::handle_t handle = bls_response_queue_->Pop();
+    if (handle == DUMMY_MESSAGE) {
+      break;
+    }
+
+    std::unique_ptr<IPCMessage> ipc_message;
+    ResponseBatch* response_batch = nullptr;
+    bi::managed_external_buffer::handle_t* response_handle = nullptr;
+    std::unique_ptr<InferResponse> infer_response;
+    bool responses_is_set = false;
+    PythonBackendException pb_exception(std::string{});
+
+    try {
+      ipc_message = IPCMessage::LoadFromSharedMemory(shm_pool_, handle);
+      AllocatedSharedMemory<char> response_batch_shm =
+          shm_pool_->Load<char>(ipc_message->Args());
+      response_batch =
+          reinterpret_cast<ResponseBatch*>(response_batch_shm.data_.get());
+      response_handle =
+          reinterpret_cast<bi::managed_external_buffer::handle_t*>(
+              response_batch_shm.data_.get() + sizeof(ResponseBatch));
+      responses_is_set = true;
+
+      if (response_batch->has_error) {
+        if (response_batch->is_error_set) {
+          std::unique_ptr<PbString> pb_string =
+              PbString::LoadFromSharedMemory(shm_pool_, response_batch->error);
+          infer_response = std::make_unique<InferResponse>(
+              std::vector<std::shared_ptr<PbTensor>>{},
+              std::make_shared<PbError>(pb_string->String()));
+        } else {
+          infer_response = std::make_unique<InferResponse>(
+              std::vector<std::shared_ptr<PbTensor>>{},
+              std::make_shared<PbError>(
+                  "An error occurred while performing BLS request."));
+        }
+      }
+
+      if (responses_is_set) {
+        // Need to acquire the GIL to avoid hangs.
+        py::gil_scoped_acquire acquire;
```

Comment:
Can we move the GIL lock to: https://github.com/triton-inference-server/python_backend/blob/0aef2c4cbee83984ce9302d4563a3c40f351d992/src/pb_tensor.cc#L470

You need to add `#ifdef TRITON_PB_STUB` around it so that it is only activated in the Python backend stub.

Reply:
I tried to add the GIL inside `pb_tensor.cc`, but somehow it failed for async cases. Need to look into it deeper. I moved the GIL to [here](https://github.com/triton-inference-server/python_backend/pull/208/files#diff-32938d76ce399e799a75fbec5050c99c8546fcc08fb75ab526ea331342731932R159) right before `PbTensor::LoadFromSharedMemory` is called. This works for async cases.